plugins/sources/manager.go

```diff
@@ -20,169 +17,238 @@
+	response               []*metrics.DataBatch
+	responseMtx            sync.Mutex
 }
 
-func (this *sourceManager) Name() string {
-	return "source_manager"
-}
+func NewEmptySourceManager() SourceManager {
+	sm := &sourceManagerImpl{
+		responseChannel:        make(chan *metrics.DataBatch),
+		metricsSourceProviders: make(map[string]metrics.MetricsSourceProvider),
+		metricsSourceTickers:   make(map[string]*time.Ticker),
+	}
+
+	sm.rotateResponse()
+	go sm.run()
 
-func (this *sourceManager) AddProvider(provider MetricsSourceProvider) {
-	this.mtx.Lock()
-	defer this.mtx.Unlock()
-	this.metricsSourceProviders[provider.Name()] = provider
-	glog.V(4).Infof("added provider: %s", provider.Name())
+	return sm
 }
 
-func (this *sourceManager) DeleteProvider(name string) {
-	this.mtx.Lock()
-	defer this.mtx.Unlock()
-	delete(this.metricsSourceProviders, name)
-	glog.V(4).Infof("deleted provider %s", name)
+// NewSourceManager creates a new NewSourceManager with the configured goMetricsSourceProviders
+func NewSourceManager(src flags.Uris) SourceManager {
+	sm := NewEmptySourceManager()
+
+	gometricsSourceProviders := buildProviders(src)
+	providerCount.Update(int64(len(gometricsSourceProviders)))
+	for _, runtime := range gometricsSourceProviders {
+		sm.AddProvider(runtime)
+	}
+
+	return sm
 }
 
-func (this *sourceManager) ScrapeMetrics(start, end time.Time) (*DataBatch, error) {
-	glog.V(1).Infof("Scraping metrics start: %s, end: %s", start, end)
+// AddProvider register and start a new goMetricsSourceProvider
+func (sm *sourceManagerImpl) AddProvider(provider metrics.MetricsSourceProvider) {
+	name := provider.Name()
+	glog.Infof("Adding provider: '%s' - collection iterval: '%v' - timeout: '%v'", name, provider.CollectionInterval(), provider.TimeOut())
+	if _, found := sm.metricsSourceProviders[name]; found {
+		glog.Fatalf("Error on 'SourceManager.AddProvider' Duplicate Metrics Source Provider name: '%s'", name)
+	}
+
+	sm.metricsSourcesMtx.Lock()
+	defer sm.metricsSourcesMtx.Unlock()
 
-	sources := []MetricsSource{}
-	this.mtx.RLock()
-	for _, sourceProvider := range this.metricsSourceProviders {
-		glog.V(2).Infof("Scraping sources from provider: %s", sourceProvider.Name())
-		sources = append(sources, sourceProvider.GetMetricsSources()...)
+	var ticker *time.Ticker
+	if provider.CollectionInterval() != 0 {
+		ticker = time.NewTicker(provider.CollectionInterval())
+	} else {
+		ticker = time.NewTicker(10 * time.Second)
+		glog.Errorf("Provider '%s' have a invalid 'CollectionInterval' using '10s", provider.Name())
 	}
-	providerCount.Update(int64(len(this.metricsSourceProviders)))
-	this.mtx.RUnlock()
 
-	sourceCount.Update(int64(len(sources)))
+	sm.metricsSourceProviders[name] = provider
+	sm.metricsSourceTickers[name] = ticker
+	glog.V(2).Infof("added provider: %s", name)
 
-	responseChannel := make(chan *DataBatch)
-	startTime := time.Now()
-	timeoutTime := startTime.Add(this.metricsScrapeTimeout)
+	go scrape(provider, sm.responseChannel)
```

Comment:
Is this line necessary?

Reply:
Not really, it was there for initial tests versions.